tensorflow_addons/metrics/kendalls_tau.py

```diff
@@ -0,0 +1,206 @@
+          y_true: actual rank values.
+          y_pred: predicted rank values.
+          sample_weight (optional): Ignored.
+
+        Returns:
+          Update op.
+        """
+        if y_true.shape and y_true.shape[0]:
+            print("update_state %s %s" % (y_true, y_pred))
+            print("update_shape %s" % (y_true.shape[0]))
+            i = tf.searchsorted(
+                self.actual_cuts,
+                tf.cast(tf.reshape(y_true, -1), self.actual_cuts.dtype),
+            )
+            j = tf.searchsorted(
+                self.preds_cuts, tf.cast(tf.reshape(y_pred, -1), self.preds_cuts.dtype)
+            )
+
+            def body(k, n, m, nrow, ncol):
+                return (
+                    k + 1,
+                    n + 1,
+                    tf.sparse.add(
+                        m,
+                        tf.SparseTensor(
+                            [[i[k], j[k]]],
+                            tf.cast([1], dtype=self.m.dtype),
+                            self.m.shape,
+                        ),
+                    ),
+                    tf.sparse.add(
+                        nrow,
+                        tf.SparseTensor(
+                            [[i[k]]],
+                            tf.cast([1], dtype=self.nrow.dtype),
+                            self.nrow.shape,
+                        ),
+                    ),
+                    tf.sparse.add(
+                        ncol,
+                        tf.SparseTensor(
+                            [[j[k]]],
+                            tf.cast([1], dtype=self.ncol.dtype),
+                            self.ncol.shape,
+                        ),
+                    ),
+                )
+
+            _, self.n, self.m, self.nrow, self.ncol = tf.while_loop(
+                lambda k, n, m, nrow, ncol: k < i.shape[0],
+                body=body,
+                loop_vars=(0, self.n, self.m, self.nrow, self.ncol),
+            )
+
+    def result(self):
+        m_dense = tf.sparse.to_dense(tf.cast(self.m, tf.float32))
+        n_cap = tf.cumsum(
+            tf.cumsum(
+                tf.slice(tf.pad(m_dense, [[1, 0], [1, 0]]), [0, 0], self.m.shape),
+                axis=0,
+            ),
+            axis=1,
+        )
+        # Number of concordant pairs.
+        p = tf.math.reduce_sum(tf.multiply(n_cap, m_dense))
+        sum_m_squard = tf.math.reduce_sum(tf.math.square(m_dense))
+        # Ties in x.
+        t = (
+            tf.math.reduce_sum(tf.math.square(tf.sparse.to_dense(self.nrow)))
+            - sum_m_squard
+        ) / 2.0
+        # Ties in y.
+        u = (
+            tf.math.reduce_sum(tf.math.square(tf.sparse.to_dense(self.ncol)))
+            - sum_m_squard
+        ) / 2.0
+        # Ties in both.
+        b = tf.math.reduce_sum(tf.multiply(m_dense, (m_dense - 1.0))) / 2.0
+        # Number of discordant pairs.
+        n = tf.cast(self.n, tf.float32)
+        q = (n - 1.0) * n / 2.0 - p - t - u - b
```

Comment:
I don't know if I lost some pass but is in 16 `n+1`?

Reply:
This is the number of ways to choose two from n items is (n-1)*n / 2, that's the total number of pairs. I think this is correct, and it wouldn't produce the right scores if it wasn't.

Reply:
It was just cause 16 is:![immagine](https://user-images.githubusercontent.com/1710528/114208101-662a7980-995d-11eb-8536-0a3cbca656bd.png)

Reply:
One based versus zero based indexing.